packages/sitecore-jss-react/src/components/BYOCComponent.tsx

```diff
@@ -119,30 +131,75 @@ export class BYOCComponent extends React.Component<BYOCComponentProps> {
 
     let componentProps: { [key: string]: unknown } = undefined;
 
-    if (props.params?.ComponentProps) {
-      try {
-        componentProps = JSON.parse(props.params.ComponentProps) ?? {};
-      } catch (e) {
-        console.warn(
-          `Parsing props for ${componentName} component from rendering params failed. Error: ${e}`
-        );
-        return this.props.errorComponent ? (
-          <this.props.errorComponent error={e as Error} />
-        ) : (
-          <DefaultErrorComponent error={e as Error} />
-        );
+    if (props.fetchedData === null) {
+      if (props.params?.ComponentProps) {
+        try {
+          componentProps = JSON.parse(props.params.ComponentProps) ?? {};
+        } catch (e) {
+          console.warn(
+            `Parsing props for ${componentName} component from rendering params failed. Error: ${e}`
+          );
+          return this.props.errorComponent ? (
+            <this.props.errorComponent error={e as Error} />
+          ) : (
+            <DefaultErrorComponent error={e as Error} />
+          );
+        }
       }
     }
-    if (!componentProps) {
+    if (!componentProps && props.fetchedData === null) {
       componentProps = props.fields ? getDataFromFields(props.fields) : {};
     }
 
+    const data = props.fetchedData || componentProps || {};
+
     return (
       <FEAAS.ExternalComponent
         componentName={componentName}
         fallback={fallbackComponent}
-        {...componentProps}
+        {...data}
       />
     );
   }
 }
+
+/**
+ * @param {BYOCComponentParams} params rendering parameters for BYOC component
+ */
+export async function fetchBYOCComponentServerProps(
+  params: BYOCComponentParams
+): Promise<BYOCComponentProps> {
+  let fetchedData: FEAAS.DataScopes = null;
+  const fetchDataOptions: FEAAS.DataOptions = params.ComponentProps
+    ? JSON.parse(params.ComponentProps)
+    : {};
+
+  try {
+    fetchedData = await fetchData(fetchDataOptions);
+  } catch (e) {
+    console.error(e);
+  }
+
+  return {
+    fetchedData,
+  };
+}
+
+/**
+ * Fetches component data based on the provided data options.
+ * This function asynchronously fetches data using the FEAAS.DataSettings.fetch method.
+ *
+ * @param {FEAAS.DataOptions} dataOptions - Options to customize data fetching.
+ * @returns {Promise<FEAAS.DataScopes>} A promise that resolves with the fetched data,
+ * or rejects with an error if data fetching encounters an issue.
+ * @throws {Error} If an error occurs during data fetching, it is propagated as an error.
+ */
+async function fetchData(dataOptions: FEAAS.DataOptions): Promise<FEAAS.DataScopes> {
```

Comment:
Do we need this function separately? Seems like DataSettings.fetch can be called in fetchBYOCComponentServerProps with same error handling logic

Reply:
Yeah i was trying to make it follow similar pattern as Feaas but this is a overkill.
Refactored.